### DISCUSSION TITLE
SQLAlchemy asyncio appears to block the event loop on ASGI web frameworks like FastAPI

### DISCUSSION THREAD

**Discussion by msukmanowsky** [fastapi/fastapi#12089] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Before running this example you'll need to create a database (filename db.sqlite3) with the following schema:
# CREATE TABLE todo (
#     id INTEGER PRIMARY KEY,
#     title TEXT NOT NULL,
#     completed_at TIMESTAMP
# );
# You should also insert ~1M rows into the table to see the difference in performance.
# The code in this repo https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness does this.
import asyncio
import datetime as dt
import typing as T

import sqlalchemy as sa
from fastapi import Depends, FastAPI
from sqlalchemy.ext.asyncio import (
    AsyncSession,
    async_sessionmaker,
    create_async_engine,
)
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, sessionmaker

# Database stuff ##############################################################

sync_engine = sa.create_engine(
    "sqlite:///./db.sqlite3",
)
sync_session_factory = sessionmaker(bind=sync_engine, expire_on_commit=False)

async_engine = create_async_engine(
    "sqlite+aiosqlite:///./db.sqlite3",
)
async_session_factory = async_sessionmaker(bind=async_engine, expire_on_commit=False)


class Base(DeclarativeBase):
    pass


class Todo(Base):
    __tablename__ = "todo"

    id: Mapped[int] = mapped_column(primary_key=True)
    title: Mapped[str]
    completed_at: Mapped[dt.datetime | None]


async def get_async_db_session():
    async with async_session_factory() as session:
        yield session


AsyncDBSessionDep = T.Annotated[AsyncSession, Depends(get_async_db_session)]


def get_sync_db_session():
    with sync_session_factory() as session:
        yield session


SyncDBSessionDep = T.Annotated[AsyncSession, Depends(get_sync_db_session)]


app = FastAPI()


@app.get("/ping")
async def ping():
    # This is a simple async route that is meant as a control case
    return {"message": "pong"}


@app.get("/ping-with-async-sleep")
async def ping_with_async_sleep():
    # This endpoint demonstrates the "right way" to do blocking async work
    await asyncio.sleep(10)
    return {"message": "pong"}


completed_at_status = sa.case(
    (Todo.completed_at.is_(None), "not_completed"),
    else_="completed",
)
count_todos_query = sa.select(
    completed_at_status.label("completed_at_status"),
    sa.func.count(Todo.id).label("count"),
).group_by(completed_at_status)


@app.get("/count-todos-async")
async def count_todos_async(db_session: AsyncDBSessionDep):
    # Since this route is defined with `async def`, it'll run in an asyncio
    # event loop. Because we're using an async database session, this *should*
    # not block the event loop, but that's not what we see.

    result = (await db_session.execute(count_todos_query)).all()
    response = {"completed": 0, "not_completed": 0}
    for row in result:
        response[row.completed_at_status] = row.count
    return response


@app.get("/count-todos-sync")
def count_todos_sync(db_session: SyncDBSessionDep):
    # Since this route is defined with `def`, it'll run in a threadpool
    # and won't block the asyncio event loop

    result = db_session.execute(count_todos_query).all()
    response = {"completed": 0, "not_completed": 0}
    for row in result:
        response[row.completed_at_status] = row.count

    return response
```

### Description

Long running queries using SQLAlchemy's asyncio support seem to block the FastAPI ASGI event loop causing all responses to queue and ultimately slow down.

I created a more in-depth example that can be run in [this repo](https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness)

At first, we believed this to be an issue with SQLALchemy blocking the event loop directly, but in [this discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/11804) they demonstrated examples where this is clearly not the case.

However, when making concurrent requests to endpoints that use `sqlalchemy.ext.asyncio.AsyncSession` (`/count-todos-async` in this example) and those that don't (e.g. `/ping` in this example), we clearly see evidence of the FastAPI event loop being blocked resulting in latencies spiking on `/ping`:

```
------ START ------
Running test with 100 requests each to paths: ['/count-todos-async', '/ping']
/count-todos-async:
Min: 6.99
Mean: 12.97
Max: 13.60
/ping:
Min: 7.05
Mean: 11.47
Max: 13.17
------ END ------
```

Notice that the `/ping` endpoint experiences latencies far exceeding a simple control case:

```
------ START ------
Running test with 100 requests each to paths: ['/ping']
/ping:
Min: 0.02
Mean: 0.04
Max: 0.05
------ END ------
```

And when using `asyncio.sleep` we see that only the `/ping-with-async-slee` endpoint is impacted, not `/ping`:
```
------ START ------
Running test with 100 requests each to paths: ['/ping', '/ping-with-async-sleep']
/ping:
Min: 0.10
Mean: 0.23
Max: 0.28
/ping-with-async-sleep:
Min: 10.24
Mean: 10.29
Max: 10.32
------ END ------
```

### Operating System

macOS

### Operating System Details

14.6.1 (23G93)

### FastAPI Version

0.112.2

### Pydantic Version

2.8.2

### Python Version

Python 3.12.5

**Comment by YuriiMotov** [comment]
Are you sure it caused by SQLAlchemy?
Try commenting `for` loops or inserting `await asyncio.sleep(0)` in the bodies of `for` loops:

```
@app.get("/count-todos-async")
async def count_todos_async(db_session: AsyncDBSessionDep):
    # Since this route is defined with `async def`, it'll run in an asyncio
    # event loop. Because we're using an async database session, this *should*
    # not block the event loop, but that's not what we see.

    result = (await db_session.execute(count_todos_query)).all()
    response = {"completed": 0, "not_completed": 0}
    for row in result:
        await asyncio.sleep(0). # make the for-loop non-bloking
        response[row.completed_at_status] = row.count
    return response

```

Python can only switch between async tasks when it meets the `await` instruction. Your `for`-loops have potentially a lot of iterations and there is no one `await` instruction in these loops. Python can't switch to other async tasks during the execution of long `for` loop. This is probably the reason why you see these latencies

**Reply by msukmanowsky** [reply]
In this case, there are two rows in the result set. That's not the cause of any latency here.

**Reply by YuriiMotov** [reply, marked as answer]
I just remembered that I had a similar problem when I was trying to imitate high load using `SQLAlchemy` with `aiosqlite` driver.
At the same time everything worked Ok when I used `postgres` driver.

As I understand the problem appears when the number of connections is more when the size of `SQLAlchemy`'s connection pool.
As I remember you can overcome it by setting `pool_size=100` (or `max_overflow=100`) for your `engine`.
